server/views.py

```diff
@@ -1,444 +1,331 @@
+    """
+    Path: /offerings/<canvas_id>/exams/<exam_name>/toggle
+    Toggles an exam for a course offering.
+    """
+    if exam.is_active:
+        exam.is_active = False
+    else:
+        # only one exam can be active at a time, so deactivate all others first
+        exam.offering.mark_all_exams_as_inactive()
+        exam.is_active = True
+    db.session.commit()
+    return redirect(url_for('offering', offering=exam.offering))
+
+
+@app.route('/<exam:exam>/')
+def exam(exam):
+    """
+    Path: /offerings/<canvas_id>/exams/<exam_name>
+    Front page for an exam, which essentially shows all rooms created for an exam.
+    """
+    return render_template('exam.html.j2', exam=exam)
+# endregion
+
+# region Room CRUDI
+
 
 @app.route('/<exam:exam>/rooms/import/')
 @google_oauth.required(scopes=['https://www.googleapis.com/auth/spreadsheets.readonly'])
 def import_room(exam):
+    """
+    Path: /offerings/<canvas_id>/exams/<exam_name>/rooms/import
+    """
     new_form = RoomForm()
-    choose_form = MultRoomForm()
-    return render_template('new_room.html.j2', exam=exam, new_form=new_form, choose_form=choose_form)
+    choose_form = ChooseRoomForm()
+    return render_template('new_room.html.j2',
+                           exam=exam, new_form=new_form, choose_form=choose_form)
 
-@app.route('/<exam:exam>/rooms/import/new/', methods=['GET', 'POST'])
+
+@app.route('/<exam:exam>/rooms/import/from_custom_sheet/', methods=['GET', 'POST'])
 @google_oauth.required(scopes=['https://www.googleapis.com/auth/spreadsheets.readonly'])
-def new_room(exam):
+def import_room_from_custom_sheet(exam):
+    """
+    Path: /offerings/<canvas_id>/exams/<exam_name>/rooms/import/new
+    """
     new_form = RoomForm()
-    choose_form = MultRoomForm()
+    choose_form = ChooseRoomForm()
     room = None
     if new_form.validate_on_submit():
         try:
-            room = validate_room(exam, new_form)
-        except ValidationError as e:
+            room = parse_form_and_validate_room(exam, new_form)
+        except Exception as e:
             new_form.sheet_url.errors.append(str(e))
         if new_form.create_room.data:
-            db.session.add(room)
-            db.session.commit()
+            try:
+                db.session.add(room)
+                db.session.commit()
+                # TODO: proper error handling
+            except:
+                new_form.sheet_url.errors.append(
+                    "Room name {} already exists for this exam.".format(room.name))
             return redirect(url_for('exam', exam=exam))
-    return render_template('new_room.html.j2', exam=exam, new_form=new_form, choose_form=choose_form, room=room)
+    return render_template('new_room.html.j2',
+                           exam=exam, new_form=new_form, choose_form=choose_form, room=room)
+
 
-@app.route('/<exam:exam>/rooms/import/choose/', methods=['GET', 'POST'])
+MASTER_ROOM_SHEET = 'https://docs.google.com/spreadsheets/d/' + \
```

Comment:
Is this in the docs? We should just put this in an app-wide config file. 
TBD is how we should update rooms / sync shared data

Reply:
yes that can be in env var